File: src/main/java/me/melijn/dhs/storage/CacheManager.java
package me.melijn.dhs.storage;

import me.melijn.dhs.components.SwitchComponent;
import me.melijn.dhs.components.User;

import java.util.ArrayList;
import java.util.List;

public class CacheManager {

    private Database database;
    public final List<User> users = new ArrayList<>();
    public final List<SwitchComponent> switchComponentList = new ArrayList<>();

    public CacheManager(Database database) {
        this.database = database;
        refreshCaches();
    }

    public void refreshCaches() {
        switchComponentList.clear();
        switchComponentList.addAll(database.getSwitchList());

        database.getUserTokens().forEach((name, token) -> {
            User user = new User(name, token);
            user.initOverrides(database);
            users.add(user);
        });
    }

    public String getUsernameFromToken(String token) {
        return users.stream().filter(user ->
                user.getToken().equals(token)
        ).findAny().map(User::getName).orElse(null);
    }

    public SwitchComponent getSwitchComponentById(int id) {
        return switchComponentList.stream().filter(switchComponent -> switchComponent.getId() == id).findAny().orElse(null);
    }
}

File: src/main/kotlin/me/melijn/dhs/database/CacheManager.kt
package me.melijn.dhs.database

import kotlinx.coroutines.runBlocking
import me.melijn.dhs.objects.components.SwitchComponent
import me.melijn.dhs.objects.components.User

class CacheManager(private val dbManager: DBManager) {

    private val users = mutableListOf<User>()
    val switchComponentList = mutableListOf<SwitchComponent>()

    suspend fun refreshCaches() {
        switchComponentList.clear()
        switchComponentList.addAll(dbManager.switchWrapper.getSwitches())
        dbManager.userWrapper.getUserTokens().forEach { (name: String, token: String) ->
            val user = User(name, token)
            user.initOverrides(dbManager)
            users.add(user)
        }
    }

    fun getUsernameFromToken(token: String): String? {
        return users.stream()
            .filter { user: User -> user.token == token }
            .findAny()
            .map<String>(User::name)
            .orElse(null)
    }

    fun getSwitchComponentById(id: Int): SwitchComponent? {
        return switchComponentList.stream()
            .filter { switchComponent: SwitchComponent -> switchComponent.id == id }
            .findAny()
            .orElse(null)
    }

    fun getUserByName(username: String): User? {
        return users.stream()
            .filter { user: User -> user.name == username }
            .findAny()
            .orElse(null)
    }

    fun getSwitchComponentByCode(code: Int): SwitchComponent? {
        return switchComponentList
            .firstOrNull { it.onCode == code || it.offCode == code }
    }

    init {
        runBlocking {
            refreshCaches()
        }
    }
}